### DISCUSSION TITLE
Whisper with Urro: Native speech recognition, speaker diarization, and word-level timestamps, all in one model

### DISCUSSION THREAD

**Discussion by urroxyz** [openai/whisper#2581] [Show and tell]
# Whisper with Urro `WHISPER + URRO`

Multilingual automatic speech recognition (ASR) with speaker segmentation (SS) / speaker diarization (SD) and word-level timestamps (WLT)

## Installation

### Latest
```shell
pip install git+https://github.com/urroxyz/whisper@v0.2.0
```

### Development
```shell
pip install git+https://github.com/urroxyz/whisper
```

## Introduction

After immense experimentation, I have discovered that, yes, Whisper *can* segment speakers and timestamp words! And I have created WHISPER + URRO to offer an easy solution therefor.

By modifying the thinking process of the OpenAI model, we can force it to delimit new speakers with symbols like hyphens (`-`) or greater-thans (`>`), or even with complete labels such as `[SPEAKER 1]` and `[SPEAKER 2]` to keep track of who is speaking and when.[^acknowledgement-1] By extracting cross-attentions and processesing them with dynamic-time warping, we can reconstruct timestamps on the word level rather than relying on occasional generated time tokens.[^acknowledgement-2]

[^acknowledgement-1]: Unique to WHISPER + URRO.
[^acknowledgement-2]: As explicitly implemented in `whisper-timestamped`, alongside other libraries, such as `openai-whisper`.

## Supported models
### Official

| Size                                          | Parameters | New-speaker segmentation | Speaker diarization | Word-level timestamps |
|-----------------------------------------------|------------|--------------------------|---------------------|-----------------------|
| tiny[^link-1]<br>tiny.en[^link-2]              | 39 M       | ✓                        | x                   | ✓                     |
| base[^link-3]<br>base.en[^link-4]              | 74 M       | ✓                        | x                   | ✓                     |
| small[^link-5]<br>small.en[^link-6]            | 244 M      | ✓                        | ✓                   | ✓                     |
| medium[^link-7]<br>medium.en[^link-8]          | 769 M      | ✓                        | ✓                   | x                     |
| large-v3[^link-11]                            | 1550 M     | ✓                        | ✓                   | x                     |
| large-v3-turbo[^link-12]                      | 809 M      | ✓                        | x                   | ✓                     |

[^link-1]: https://huggingface.co/onnx-community/whisper-tiny_timestamped
[^link-2]: https://huggingface.co/onnx-community/whisper-tiny.en_timestamped
[^link-3]: https://huggingface.co/onnx-community/whisper-base_timestamped
[^link-4]: https://huggingface.co/onnx-community/whisper-base.en_timestamped
[^link-5]: https://huggingface.co/onnx-community/whisper-small_timestamped
[^link-6]: https://huggingface.co/onnx-community/whisper-small.en_timestamped
[^link-7]: https://huggingface.co/onnx-community/whisper-medium-ONNX
[^link-8]: https://huggingface.co/onnx-community/whisper-medium.en-ONNX  
[^link-11]: https://huggingface.co/onnx-community/whisper-large-v3-ONNX
[^link-12]: https://huggingface.co/onnx-community/whisper-large-v3-turbo_timestamped

### Third-party
| Model                   | Parameters | New-speaker segmentation | Speaker diarization | Word-level timestamps |
|-------------------------|------------|--------------------------|---------------------|-----------------------|
| whisper-d-v1a[^link-13] | 1550 M     | ✓                        | ✓                   | x                     |

[^link-13]: [onnx-community/whisper-d-v1a-ONNX](https://huggingface.co/onnx-community/whisper-d-v1a-ONNX)

## Comparison
| Source                                                                                                                             | Transcript                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
|------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Ground truth**                                                                                                                 | [SPEAKER 1] Down in front. <br><br> [SPEAKER 2] Hey, sit down, that’s wrong of you. <br><br> [SPEAKER 1] The little lady who is to become Mrs. Harvey Yates over my dead body. <br><br> [SPEAKER 3] I know I have the sincere wishes of all my friends… <br><br> and can only tell you how much I appreciate it. <br><br> I think I can honestly say this is the happiest moment of my life. <br><br> Look what I have here. <br><br> It’s a little engagement present just given me by Mr. Yates. |
| **Pretrained model (`medium`)**<br><br>***No speaker labels***                                                                      | Down in front. <br><br> Hey, sit down, ~~that’s fine~~. <br><br> The little lady who is to become Mrs. Harvey Yates over my dead body. <br><br> [APPLAUSE] <br><br> I know I have the sincere wishes of all my friends, <br><br> and can only tell you how much I appreciate it. <br><br> I think I can honestly say this is the happiest moment of my life. <br><br> Look what I have here… <br><br> It’s a little engagement present just given me by Mr. Yates. |
| **Pretrained model (`medium`)**<br>*with WHISPER + URRO*<br><br>`delimiter=SPEAKER()`<br>`prompt=SPEAKERS(3, "en")`<br><br>***Correct speaker labels*** | [SPEAKER 1] Down in front. <br><br> [SPEAKER 2] Hey, sit down, ~~that’s fine~~. <br><br> [SPEAKER 1] The little lady who is to become Mrs. Harvey Yates over my dead body. <br><br> [APPLAUSE] <br><br> [SPEAKER 3] I know I have the sincere wishes of all my friends, <br><br> and can only tell you how much I appreciate it. <br><br> I think I can honestly say this is the happiest moment of my life. <br><br> Look what I have here… <br><br> It’s a little engagement present just given me by Mr. Yates. |
| **Finetuned model (`d-v1a`)**<br><br>***Incorrect speaker labels***                                                               | [S1] Down in front. <br><br> [S2] Hey, sit down, ~~it’s warm~~. <br><br> [S1] The little lady who is to become Mrs. Harvey Yates, over my dead body. <br><br> ~~[S2]~~ I know I have the sincere wishes of all my friends, <br><br> and can only tell you how much I appreciate it. <br><br> I think I can honestly say this is the happiest moment of my life. <br><br> Look what I have here. |
| **Finetuned model (`d-v1a`)**<br>*with WHISPER + URRO*<br><br>`delimiter=SPEAKER(short=True)`<br>`prompt=SPEAKERS(3, "en", short=True)`<br><br>***Correct speaker labels***   | [S1] Down in front. <br><br> [S2] Hey, sit down, ~~it’s warm~~. <br><br> [S1] The little lady who is to become Mrs. Harvey Yates, over my dead body. <br><br> [S3] I know I have the sincere wishes of all my friends, <br><br> and can only tell you how much I appreciate it. <br><br> I think I can honestly say this is the happiest moment of my life. <br><br> Look what I have here. |

**Comment by kinchitagg** [comment]
Hey this looks interesting to me. The speaker segregation is an important aspect of generating subtitle as it can give clues for video editing via automation. However, is your project support api if I ran it locally ?  Also to test your project should I just install your package ? can you help ?

**Reply by urroxyz** [reply]
Thank you so much for your interest.
Please visit [the repo page](https://whisper.urro.xyz/) and review its README.md for more information.

**Comment by TheJeje20** [comment]
Hi,
It looks very interesting to me too, but I can't find any trace of the project and your website is offline. 
Is it possible to make a repo available back online?

Many thanks for sharing

**Reply by urroxyz** [reply]
Thanks for your interest.
I privated the repo a while ago because I stopped maintaining it, didn't see much activity, and it had also stopped working in some scenarios. Didn't want to disappoint anyone.

Its visibility is public again if you want to mess around with it: [urroxyz/whisper](https://github.com/urroxyz/whisper) :)

**Reply by TheJeje20** [reply]
> Thanks for your interest. I privated the repo a while ago because I stopped maintaining it, didn't see much activity, and it had also stopped working in some scenarios. Didn't want to disappoint anyone.
> 
> Its visibility is public again if you want to mess around with it: [urroxyz/whisper](https://github.com/urroxyz/whisper) :)

Thanks a lot for sharing! I'll have a look when I have a few spare hours, I look forward to tinker with this too :)
